Python/Product/Analysis/AnalysisHashSet.cs

```diff
@@ -377,35 +432,38 @@ public AnalysisHashSet(IEnumerable<AnalysisValue> enumerable, IEqualityComparer<
                     }
                     // merging values has changed the one we should store, so
                     // replace it.
-                    var newHc = _comparer.GetHashCode(newKey) & Int32.MaxValue;
-                    if (newHc != buckets[index].HashCode) {
+                    var newHc = comparer.GetHashCode(newKey) & Int32.MaxValue;
+                    if (newHc != buckets.Buckets[index].HashCode) {
                         // The hash code should not change, but if it does, we
                         // need to keep things consistent
                         Debug.Fail("Hash code changed when merging AnalysisValues");
                     }
                     Thread.MemoryBarrier();
-                    buckets[index].Key = _removed;
-                    _count--;
-                    return AddOne(buckets, newKey, newHc);
+                    buckets.Buckets[index].Key = _removed;
+                    buckets.Count -= 1;
+                    return AddOne(ref buckets, newKey, newHc, comparer);
                 }
 
-                index = ProbeNext(buckets, index);
+                index = ProbeNext(buckets.Buckets, index);
 
                 if (index == startIndex) {
                     break;
                 }
             }
 
-            if (buckets[addIndex].Key != null &&
-                buckets[addIndex].Key != _removed &&
-                !buckets[addIndex].Key.IsAlive) {
-                _count--;
+            if (buckets.Buckets[addIndex].Key == null || buckets.Buckets[addIndex].Key == _removed) {
+                // Removal has been counted already
+                buckets.Count += 1;
+            } else if (!buckets.Buckets[addIndex].Key.IsAlive) {
+                // Remove/add means no change te count
+            } else {
+                buckets.Count += 1;
             }
-            buckets[addIndex].HashCode = hc;
+            buckets.Buckets[addIndex].HashCode = hc;
```

Comment:
What's the point of calling `Thread.MemoryBarrier();` after this line?

Reply:
According to the comment two lines below (contributed by @DinoV), it's to make sure that the hashcode is stored before the key changes.

Reply:
Then `Thread.MemoryBarrier` must be called before every call to `HashCode` field, otherwise it is useless.

Reply:
Presumably between reads of `Key` and reads of `HashCode`, yes? So that we ensure that we always read `Key` first and always write to `Key` last?

Reply:
To ensure that we read `Key` first and always write to `Key` last we need several calls:

`AddOne`:
```cs
bucket.HashCode = hashCode;
Thread.MemoryBarrier();
bucket.Key = key;
Thread.MemoryBarrier();
```

`Contains`:
```cs
Thread.MemoryBarrier();
var key = bucket.Key;
if (key != null) {
    Thread.MemoryBarrier();
    var hashCode = buckets[index].HashCode;
    if (Object.ReferenceEquals(key, existingKey) ||
        (existingKey != _removed && hashCode == hc && _comparer.Equals(key, (AnalysisValue)existingKey))) {
        return index;
    }
}
```

And that is correct if we have these two methods.

Reply:
But let's not change it right now.